### DISCUSSION TITLE
SkyWalking OAP server v9 Core

### DISCUSSION THREAD

**Discussion by wu-sheng** [apache/skywalking#8241] [Ideas]
# SkyWalking OAP server v9 Core

After the SkyWalking OAP 8.9.0 released, which is the last feature version of 8.x, I am glad to introduce the v9 core, including the new concept and plans.

## Why v9 is being introduced?
SkyWalking core concepts have been designed back to 2018, which are service, instance, endpoint, with dependencies and topology amon them.
This is the key value of the whole SkyWalking ecosytem, and also why we are successful.
With 3 years, we have evoluted from an agent-based tracing minitoring system to a full-stack(VM to agent, covering logs/trace/metrics/events).
But meaningwhile, we noticed with more levels of infrastructure being added, such as vm-monitoring, k8s monitoring, service mesh observability, incoming serverless(FAAS),
our services get mixed, and unclear about the boundaries. Also, we have plans to adopt more ways to obserse, most importantly **ebpf**.

So, a new core with new concepts must be introduced

## New concept
The most important concept in v9 core is **LAYER**. Layer represents an abstract framework in the computer science, such as operation system(VM layer), Kubernetes(k8s layer), Service Mesh(typical Isto+Envoy layer)
This kind of layer would be owners of different services detected from different technology. 
In the older 8.x branch, we have a `group` concept added to resolve this kind of mixing issue temporarily, but clearly, a new `layer` concept is much better. 
Also, `group` concept is going to be kept, as it is important inside each `layer`. At this time, group would be designed for end users to group their service internal. With `no group`, it would be in the default group.

The name of layer is a string, but we would reserve some for visualization(UI)
- os-linux
- k8s
- general(agent-installed)
- faas
- mesh
- database
- cache
(SkyWalking community holds all rights to add new reserved layer name without extra notification)

### Service
Service should still be treated as unique globally like today, even across `layer`. Layer is an attachment attribute, so it wouldn't be added in the name.
The layer of a service registered in the first time, it would be saved, but never would be overrided in the future(unless the TTL passed, it goes the 2nd time registration).

### Query Protocol
GraphQL metadata.graphqls would be expired, a new v2 would be designed by myself as the next step of this proposal.
Generally, all query(s) are according to service's layer, but would not do query to different kinds of service(Service, Browser, Database). All legacy query to Service, Browser, Database would be bridged into `layer`-based query.

### How to Determine Layer
All the engine(OAL, MAL, LAL) should provide core level mechanis to make receivers or analysis scripts could set `layer`. The default layer is `general`, which is like old-days `agent-installed` services.
Agent analysis(traces) would add `general` layer through OAL engine.
Linux monitoring would use `MAL` engine exposed(TODO) way in MAL script to set `os-linux` as the layer.

## Visualization

With this new concept adding, the visualization clearly would face big evolution too. @Fine0830 and I are going to propose a new UI implementation, named `skywalking-booster`(repo name), to the whole community.
In there, we have all monitoring type(layer) of objects as top level menu, tracing/metrics/logs would be a second level menu, and only be showing up if SkyWalking has upstream adopted features.

Of course, consider requirements of custom metrics, we are still providing custom setup mechanism like we did in RocketBot.

Meanwhile, I want to notify DaoCloud team, PMC member @TinyAllen and @JaredTan95, rocketbot UI(https://github.com/apache/skywalking-rocketbot-ui) is going to be expired too with the first v9 releasing.
`skywalking-booster` could become the new recommended UI for the next generation. And I would like to express sincere thanks be behalf of the whole SkyWalking community.
RocketBot UI helps the project to reach today's level. And from tech perspective, some key features, like topology map, is going to be moved into booster UI.

## Timeline

In SkyWalking iteration, we don't provide the timeline about when this version is planned to release official. We are going to keep this tradition.
So, within the first 9.0.0 iteration, if some critical issues(serious bugs and/or CVEs) are reported/fixed, PMC could cherry-pick commits or accept patches on 8.9.x branch, and initialize 8.9.x release.
Notice, we should have no plan to keep features being added into 8.9.x.

FYI @apache/skywalking-committers

**Comment by FeynmanZhou** [comment]
The SkyWalking 9.0.0 looks to be a full-stack APM system, the design of `layer` is interesting. It's glad to see the layer **FaaS** will be added in its design. I am curious what will be included in the **FaaS** layer? I assume the capability of tracing, topology, and metrics for functions will be basically supported at first as I see the related issues in [openfunction functions-framework](https://github.com/OpenFunction/functions-framework/issues/9) and [integration proposal](https://github.com/OpenFunction/OpenFunction/issues). If so, we could collaborate to make it happen.

**Reply by wu-sheng** [reply]
Yes, those would be added. And in general, as FaaS use Function as the basic unit, rather than endpoint in service, so, the metrics and layout could be slight different with general service and meshed service. 
I will ping you and OpenFunction team when we are going to land this on the UI and analysis part. You could consider this is a special offer for FaaS, and provide metrics all by following OpenFunction team's idea. I hope we can catch up the time, and make this available in the first 9.0 release.

**Reply by wu-sheng** [reply]
@FeynmanZhou A screenshot of new dashboard

![image](https://user-images.githubusercontent.com/5441976/161167437-7a59dcfb-24cd-4f0b-ab0e-2a8ad5a81839.png)

@arugal is polishing and will do several fixes to complete this.

We will add a link on this dashboard to OpenFunction document page site.

**Reply by FeynmanZhou** [reply]
@wu-sheng @arugal 
Great work! The new dashboard seems much neater than before.
Currently, we have an informal version of this [document](https://github.com/OpenFunction/samples/tree/main/functions/tracing). We are preparing it on the OpenFunction website, will inform it here when this document is available on the website soon.

cc @benjaminhuo @tpiperatgod @webup @felixnoo

**Reply by wu-sheng** [reply]
@FeynmanZhou What is ETA? I think SkyWalking v9 is closing to release on sometime in next week

**Reply by Felixnoo** [reply]
> @wu-sheng @arugal Great work! The new dashboard seems much neater than before. Currently, we have an informal version of this [document](https://github.com/OpenFunction/samples/tree/main/functions/tracing). We are preparing it on the OpenFunction website, will inform it here when this document is available on the website soon.
> 
> cc @benjaminhuo @tpiperatgod @webup @Felixnoo

Hi @FeynmanZhou, I'll try to improve the informal version and put it to the website ASAP.

**Comment by wu-sheng** [comment]
The general plans of tasks are as 
1. @wu-sheng leads on new protocols and design
2. @wankai123 leads on backend adjustments
3. @kezhenxu94 assists on ElasticSearch/OpenSearch storage implementation
4. @dmsolr assists on InfluxDB storage implementation
5. @LIU-WEI-git assists on IoTDB storage implementation
6. @Fine0830 leads on new booster UI. New UI repository(https://github.com/apache/skywalking-booster-ui)
7. @arugal leads on FaaS side integration, collaborate with @FeynmanZhou

**Comment by hanahmily** [comment]
I'm glad to see this happen. @lujiajing1126 is on the integration between OAP and BanyanDB primitive version. Since 9.0 is a long-term plan, I select 8.9.x as the integrating target. FWIK, the new layer concept, has a  limited impact on the storage module. We could upgrade the BanyanDB storage implementation once we have made most backend adjustments. Does it make sense?

**Reply by wu-sheng** [reply]
For storage perspective, this `layer` is a new field, and you may need another MetadataDAO implementation with the latest GraphQL.
Others are sealed by OAL modularized kernel, which should not concern your part .

**Comment by wu-sheng** [comment]
The v9 core's query APIs are proposed here, https://github.com/apache/skywalking-query-protocol/pull/63. Feedback is welcome.

**Comment by wu-sheng** [comment]
The core update is raised by @wankai123 through https://github.com/apache/skywalking/pull/8367

**Comment by wu-sheng** [comment]
The backend and metadata query v2 are ready in the master branch. The final step is waiting for [booster UI](https://github.com/apache/skywalking-booster-ui) ready, which is led by @Fine0830

**Reply by wu-sheng** [reply]
FYI @apache/skywalking-committers Feel free to use the latest codes to verify compatibility, the expectation is fully supported for all 8.9.1 compatible native and 3rd party agents.

**Comment by wu-sheng** [comment]
Closing this update. The submodule has been replaced, booster UI is the main stream now.

**Comment by wu-sheng** [comment]
SkyWalking v9 is going to release in the 2nd week, of April.

https://lists.apache.org/thread/jgpssrld6dq1wxpn49rz00gjjzjm58jm

**Reply by BFergerson** [reply]
Eager to see the new UI. I've cloned and installed the latest version of SkyWalking and the booster UI, but it looks like it's still loading the placeholder dashboard templates:
![image](https://user-images.githubusercontent.com/3278877/161536709-460fd7c7-4707-47f4-9b51-d96c0c31dfcc.png)

Is there still some work needed before it loads the real dashboard templates?

**Reply by wu-sheng** [reply]
No, there isn't. Is there a new database? It has conflicts with v8 versions.

**Reply by wu-sheng** [reply]
Yours look like the old UI version. demo.skywalking.apache.org has updated once, @kezhenxu94 will update again. You could take a look.

**Reply by BFergerson** [reply]
You're right. I was loading an older version of the UI. Looking really good so far.

Btw, I can't access https://demo.skywalking.apache.org. It times out.

**Reply by wu-sheng** [reply]
No `https`. Try `http`.
